### PR TITLE
solutions for chapter 4

### PR DESCRIPTION
**Pull request by Jeffrey04**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -629,7 +662,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = (a >>= (\x -> pure (x &&))) <*> b
```
Comment (Jeffrey04) [inline src/Chapter4.hs:665]: This fails when `a = Just False` and `b = Nothing`
shouldn't all computation involving `Nothing` returns `Nothing`?

Comment (Jeffrey04) [reply src/Chapter4.hs:665]: also tried with
```
andM a b = do
  m<-a
  n<-b
  return (m && n)
```
and simpler
```
fmap (&&) a <*>
```
still failing the case

Comment (Jeffrey04) [reply src/Chapter4.hs]: kinda fixed in the latest revision, not sure if the solution is correct tho

File: src/Chapter4.hs [hunk 2]
```diff
@@ -259,6 +267,7 @@ name.
 
 > QUESTION: Can you understand why the following implementation of the
   Functor instance for Maybe doesn't compile?
+  ANSWER: there is no transformation from (a) to (f a) hence violates the signature
```
Comment (chshersh) [inline src/Chapter4.hs:270]: That's correct 👍🏻 In other words, variable `x` has type `Maybe a`, so you can return it because you need to return a value of type `Maybe b`. The compiler is not smart enough to guess what you mean here, so you need to pattern match on the existing value and return a different value using the `Nothing` constructor.

File: src/Chapter4.hs [hunk 3]
```diff
@@ -472,10 +486,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
```
Comment (chshersh) [inline src/Chapter4.hs:489]: Great eta-reduced implementation! 💎

File: src/Chapter4.hs [hunk 4]
```diff
@@ -472,10 +486,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
...
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) (Reward f) m1 = fmap f m1
```
Comment (chshersh) [inline src/Chapter4.hs:493]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 5]
```diff
@@ -488,7 +503,22 @@ Implement the 'Applicative' instance for our 'List' type.
...
+
+  (<*>) Empty _ = Empty
+  (<*>) fs ms =
+    goM Empty ms
+    where
+      goF acc Empty _ = acc
+      goF acc (Cons f fTail) m = goF (Cons (f m) acc) fTail m
 
+      goM acc Empty = acc
+      goM acc (Cons m mTail) = goF (goM acc mTail) fs m
```
Comment (chshersh) [inline src/Chapter4.hs:519]: I would strongly advise adding type signatures to local functions (`goF` and `goM` in this case). For example, after adding a type signature you may notice some similarities with existing functions and probably assign clearer names 🙂 Current implementation is a bit difficult to read because of existing names.

If you have only a single recursive function in `where`, it's okay to call it `go`, as you already have the name for the top-level function, so no need to create an extra name for recursive helper.

File: src/Chapter4.hs [hunk 6]
```diff
@@ -629,7 +659,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = a >>= (\x -> if x then b >>= (\y -> pure (x && y)) else pure x)
```
Comment (chshersh) [inline src/Chapter4.hs:662]: That's correct implementation! Though, it could be simplified a bit, knowing the fact that `True && x` is always `x`.

```suggestion
andM a b = a >>= (\x -> if x then b else pure False)
```

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

File: src/Chapter4.hs (hunk 1 above)
Comment (chshersh) [reply src/Chapter4.hs]: If you don't use the `>>=` operator (like in the last implementation), then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.

The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face:

**Comment by Jeffrey04** [comment]
> tada Fantastic work!
> Thank you so much for taking this course! I hope you had fun man_dancing microphone
> 
> We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!
> 
> You can also ask any questions and we can discuss any topics you would like to dig into as well slightly_smiling_face Don't be a stranger!

thanks for the feedback, it was overall a great fun experience (it is nice to have a mentor to talk to while going through the materials, especially on types(Chapter 3) and Monad (Chapter 4) where I struggle the most)

File: src/Chapter4.hs (hunk 3 above)
Comment (Jeffrey04) [reply src/Chapter4.hs:489]: (:

File: src/Chapter4.hs (hunk 4 above)
Comment (Jeffrey04) [reply src/Chapter4.hs:493]: (:

File: src/Chapter4.hs (hunk 6 above)
Comment (Jeffrey04) [reply src/Chapter4.hs:662]: AH!!!!! thanks (:

File: src/Chapter4.hs (hunk 2 above)
Comment (Jeffrey04) [reply src/Chapter4.hs:270]: (:

File: src/Chapter4.hs (hunk 5 above)
Comment (Jeffrey04) [reply src/Chapter4.hs]: I suppose one should do `flatten` another `map` lol, but can't find a way to make it

File: src/Chapter4.hs (hunk 1 above)
Comment (Jeffrey04) [reply src/Chapter4.hs]: got it, thanks for the explanation (:

File: src/Chapter4.hs [hunk 7]
```diff
@@ -508,15 +508,17 @@ instance Applicative List where
...
-      goF acc (Cons f fTail) m = goF (Cons (f m) acc) fTail m
+      loopM :: List b -> List a -> List (a -> b) -> List b
+      loopM acc Empty _ = acc
+      loopM acc (Cons mHead mTail) f = loopF (loopM acc mTail f) f mHead
 
-      goM acc Empty = acc
-      goM acc (Cons m mTail) = goF (goM acc mTail) fs m
+      loopF :: List b -> List (a -> b) -> a -> List b
+      loopF acc Empty _ = acc
+      loopF acc (Cons f fTail) m = Cons (f m) $ loopF acc fTail m
```
Comment (chshersh) [inline src/Chapter4.hs]: Now it's a bit cleaner what's going on 🙂 So, my suggestion to simplify the implementation is the following:

1. In the `<*>` operator, handle two cases with `Empty` separately: `_ <*> Empty` and `Empty <*> _`.
2. After that, you will have a third case where the first pattern is `Cons`: `Cons f fs <*> xs`.
3. In that third case, you can apply the `f` function to `xs` and append to the result of `fs <*> xs`. For this, you need to implement a function for appending two lists: `append :: List a -> List a -> List a`.

I believe, this new final implementation will be cleaner and simpler 🙂

Comment (Jeffrey04) [reply src/Chapter4.hs]: hmm, let me try again

File: src/Chapter4.hs [hunk 8]
```diff
@@ -504,21 +504,22 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 {- ORMOLU_ENABLE -}
+-- copy pasta from chap3 instead of importing
+append :: List a -> List a -> List a
+append x y = merge y x
+  where
+    merge :: List a -> List a -> List a
+    merge acc Empty = acc
```
Comment (vrom911) [inline src/Chapter4.hs:512]: You can add one more case to make your solution even more efficient:

```haskell
    merge Empty xs = xs
```

In that case, you won't need to walk through each element of the second list when the first one is empty and return the list instantly 🚤

Also, you can inline merge function directly into the append, if you will switch the arguments in the pattern matching 🙂

File: src/Chapter4.hs [hunk 9]
```diff
@@ -504,21 +504,22 @@ Implement the 'Applicative' instance for our 'List' type.
...
-  (<*>) fs ms = loopM Empty ms fs
-    where
-      loopM :: List b -> List a -> List (a -> b) -> List b
-      loopM acc Empty _ = acc
-      loopM acc (Cons mHead mTail) f = loopF (loopM acc mTail f) f mHead
-
-      loopF :: List b -> List (a -> b) -> a -> List b
-      loopF acc Empty _ = acc
-      loopF acc (Cons f fTail) m = Cons (f m) $ loopF acc fTail m
+  (<*>) (Cons f fs) ms = append (f <$> ms) (fs <*> ms)
```
Comment (vrom911) [inline src/Chapter4.hs:522]: This is a really fantastic solution 👍🏼

Comment (Jeffrey04) [reply src/Chapter4.hs:522]: Thanks

**Comment by Jeffrey04** [comment]
Just an additional question regarding this, is the "Kinds" mentioned in this chapter related to higher kinded type? Do you have more introductory guides/text u think that would be helpful if I want to learn more about it?

File: src/Chapter4.hs (hunk 8 above)
Comment (Jeffrey04) [reply src/Chapter4.hs]: OOOOOOOOOOOooooOOOOOOO, thanks

**Comment by chshersh** [comment]
@Jeffrey04 Yes, kinds are very related to higher-kinded types. Similar to higher-order functions that take other functions as arguments, higher-kinded types take other type constructors as arguments. The `IntBox` type from this chapter is an example of such higher-kinded type.

You can check the "Pure Gold by ImpurePics" section in the first issue of @bindthegap FP magazine, to get an overview of kinds:

* https://bindthegap.news/issues/01nov2020.html

The following blog post is also a good introduction to kinds in Haskell:

* https://diogocastro.com/blog/2018/10/17/haskells-kind-system-a-primer/

**Comment by Jeffrey04** [comment]
thanks for the info (:

**Comment by vrom911** [comment]
Thanks again for taking the course! 🙌🏼 
I am going to close this PR, but feel free to reach out to us if you have any questions, don't be a stranger! 🤗 

And Merry Christmas 🎄 ❄️

**Comment by Jeffrey04** [comment]
thanks for all the help and guidance
